r/src/altrep.cpp

```diff
@@ -366,6 +388,277 @@ struct AltrepVectorPrimitive : public AltrepVectorBase<AltrepVectorPrimitive<sex
+
+    if (WasUnified(alt)) {
+      int j = 0;
+
+      // find out which is the first chunk of the chunk array
+      // that is present in the slice, because the main loop
+      // needs to refer to the correct transpose buffers
+      int64_t k = 0;
+      for (; j < chunked_array->num_chunks(); j++) {
+        auto nj = chunked_array->chunk(j)->length();
+        if (k + nj > start) {
+          break;
+        }
+
+        k += nj;
+      }
+
+      int* out = buf;
+      for (const auto& array : slice->chunks()) {
+        const auto& indices =
+            internal::checked_cast<const DictionaryArray&>(*array).indices();
+
+        // using the transpose data for this chunk
+        const auto* transpose_data =
+            reinterpret_cast<const int32_t*>(GetArrayTransposed(alt, j)->data());
+        auto transpose = [transpose_data](int x) { return transpose_data[x]; };
+
+        GetRegionDispatch(array, indices, transpose, out);
+
+        out += array->length();
+        j++;
+      }
+
+    } else {
+      // simpler case, identity transpose
+      auto transpose = [](int x) { return x; };
+
+      int* out = buf;
+      for (const auto& array : slice->chunks()) {
+        const auto& indices =
+            internal::checked_cast<const DictionaryArray&>(*array).indices();
+
+        GetRegionDispatch(array, indices, transpose, out);
+
+        out += array->length();
+      }
+    }
+
+    return slice->length();
+  }
+
+  template <typename Transpose>
+  static void GetRegionDispatch(const std::shared_ptr<Array>& array,
+                                const std::shared_ptr<Array>& indices,
+                                Transpose transpose, int* out) {
+    switch (indices->type_id()) {
+      case Type::UINT8:
+        GetRegionTranspose<UInt8Type>(array, indices, transpose, out);
+        break;
+      case Type::INT8:
+        GetRegionTranspose<Int8Type>(array, indices, transpose, out);
+        break;
+      case Type::UINT16:
+        GetRegionTranspose<UInt16Type>(array, indices, transpose, out);
+        break;
+      case Type::INT16:
+        GetRegionTranspose<Int16Type>(array, indices, transpose, out);
+        break;
+      case Type::INT32:
+        GetRegionTranspose<Int32Type>(array, indices, transpose, out);
+        break;
+      case Type::UINT32:
+        GetRegionTranspose<Int32Type>(array, indices, transpose, out);
+        break;
+      default:
+        break;
+    }
+  }
+
+  template <typename Type, typename Transpose>
+  static void GetRegionTranspose(const std::shared_ptr<Array>& array,
+                                 const std::shared_ptr<Array>& indices,
+                                 Transpose transpose, int* out) {
+    using index_type = typename arrow::TypeTraits<Type>::ArrayType::value_type;
+    auto raw_indices = indices->data()->GetValues<index_type>(1);
+
+    auto n = array->length();
+
+    // then set the R NA sentinels if needed
+    if (indices->null_count() > 0) {
```

Comment:
Can use `VisitArrayDataInline` here:
```c++
VisitArrayDataInline<Type>(*array->data(),
  /*valid_func=*/ [&](index_type index) {
    *out++ = transpose(index) + 1;
  },
  /*null_func=*/ [&]() {
    *out++ = cpp11::na<int>();
  });
```

Reply:
Nice. We can probably use this in other places.